src/circuit.rs

```diff
@@ -0,0 +1,467 @@
+use std::iter;
+
+use rand::Rng;
+
+use crate::{
+    Delta, EvaluatedWire, GarbledWire, GarbledWires, Gate, GateError, S, WireError, WireId,
+    core::{gate::CorrectnessError, gate_type::GateCount},
+};
+
+type DefaultHasher = blake3::Hasher;
+
+/// Errors that can occur during circuit operations
+#[derive(Debug, Clone, thiserror::Error, PartialEq, Eq)]
+pub enum CircuitError {
+    /// Gate-related error
+    #[error("Gate error: {0}")]
+    Gate(#[from] GateError),
+    /// Circuit garbling failed
+    #[error("Garbling failed: {0}")]
+    GarblingFailed(String),
+}
+
+#[derive(Clone, Debug)]
+pub struct Circuit {
+    pub num_wire: usize,
+    pub input_wires: Vec<WireId>,
+    pub output_wires: Vec<WireId>,
+    pub gates: Vec<Gate>,
+    pub gate_count: GateCount,
+}
+
+impl Default for Circuit {
+    fn default() -> Self {
+        Self {
+            num_wire: 2,
+            input_wires: Default::default(),
+            output_wires: Default::default(),
+            gates: Default::default(),
+            gate_count: GateCount::default(),
+        }
+    }
+}
+
+impl Circuit {
+    pub fn get_false_wire_constant(&self) -> WireId {
+        WireId(0)
+    }
+
+    pub fn get_true_wire_constant(&self) -> WireId {
+        WireId(1)
+    }
+
+    pub fn issue_wire(&mut self) -> WireId {
+        let new = WireId(self.num_wire);
+        self.num_wire += 1;
+        new
+    }
+
+    pub fn issue_input_wire(&mut self) -> WireId {
+        let wire_id = self.issue_wire();
+        self.make_wire_input(wire_id);
+        wire_id
+    }
+
+    pub fn issue_output_wire(&mut self) -> WireId {
+        let wire_id = self.issue_wire();
+        self.make_wire_output(wire_id);
+        wire_id
+    }
+
+    pub fn make_wire_input(&mut self, w: WireId) {
+        match self.input_wires.binary_search(&w) {
+            Ok(_) => {}
+            Err(pos) => self.input_wires.insert(pos, w),
+        }
+    }
+
+    pub fn make_wire_output(&mut self, w: WireId) {
+        match self.output_wires.binary_search(&w) {
+            Ok(_) => {}
+            Err(pos) => self.output_wires.insert(pos, w),
+        }
+    }
+
+    pub fn add_gate(&mut self, gate: Gate) {
+        self.gate_count.handle(gate.gate_type);
```

Comment:
I would suggest we use tracing and tracing_subscriber to trace the gate metrics, this can trace both the time and gate number. 

For example:


fn evaluate_gate(gate_number: usize, gate_type: GateType, a: bool, b: Option<bool>) -> bool {
    let span = span!(
        Level::DEBUG,
        "gate",
        gate_number,
        gate_type = %gate_type
    );
    let _enter = span.enter();

    let result = match gate_type {
        GateType::And => a & b.unwrap(),
        GateType::Or => a | b.unwrap(),
        GateType::Xor => a ^ b.unwrap(),
        GateType::Not => !a,
    };

    info!(%result, "Gate evaluated");

    result
}

Reply:
We can, but since we have critically many operations, I want to add tracing+spans as a separate PR and try to evaluate its impact on runtime.

Based on tracins-spans, it is trivial to make an automatic infra-structure to track the lifetime of individual functions, gates and all of that